### ISSUE TITLE
SophiaH implementation is not correct

### ISSUE THREAD

**Issue by Vectorrent** [kozistr/pytorch_optimizer#278]
## Describe the bug

There is a serious memory leak in SophiaH, making it completely unusable in nearly all real-world scenarios.

## To Reproduce

* OS : Linux
* PyTorch version : 2.4.1
* Python version : 3.12.6
* pytorch_optimizer version: 3.1.2
* reproducible codes : You can find my debug code [here](https://github.com/0-5788719150923125/praxis/blob/sophiah-debug/run.py#L600). Instructions are in the README.

## Log

There are no logs to speak of, besides the standard CUDA out-of-memory errors. There is also this warning:
```
/home/crow/repos/praxis/venv/lib/python3.12/site-packages/torch/autograd/graph.py:768: UserWarning: Using backward() with create_graph=True will create a reference cycle between the parameter and its gradient which can cause a memory leak. We recommend using autograd.grad when creating the graph to avoid this. If you have to use this function, make sure to reset the .grad fields of your parameters to None after use to break the cycle and avoid the leak. (Triggered internally at ../torch/csrc/autograd/engine.cpp:1203.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
```

## Expected behavior

I would expect SophiaH memory usage to be consistent with the [original research publication](https://arxiv.org/abs/2305.14342). From the paper:

_"Sophia maintains almost the memory and average time per step [as Adam]"_ 

On my machine, a 23M parameter language model, with a batch_size of 16 and a seq_len of 512 requires 4GB of VRAM, when using AdamW.

When using SophiaH, a 10M parameter language model, with a batch_size of 16 and a seq_len of 512 - SophiaH requires 8GB of VRAM - and crashes before the first Hessian update.

Any attempt to use gradient accumulation will cause the VRAM to "stack" - essentially doubling for every accumulation step, until CUDA crashes.

## Additional context

[The official SophiaG implemenation](https://github.com/Liuhong99/Sophia/tree/main) DOES NOT have this problem (and in fact, uses less memory than AdamW). [Your example code](https://github.com/kozistr/pytorch_optimizer/issues/194#issuecomment-1723167466) does not have this problem, either; even after scaling your example code up to 50M parameters, SophiaH does not even use 1GB of VRAM.

The main difference between your code and my code has to do with the complexity of the model. Your example model is extremely, extremely simple, while mine is complex - with **extremely complex gradients**. I believe this is where the true problem lies.

Parameter count is not the only contributor to VRAM here. The complexity of the gradients, the number of activations, etc - all of that is contributing to VRAM requirements, and causing your implementation of SophiaH to require at least 15x more VRAM than AdamW.

This is exactly the same kind of problem I was running into ~1.5 years ago, when I tried SophiaH for the first time, with a completely different code base. I don't think the problem is with my code; I think there is something wrong with this version of the optimizer.

**Comment by kozistr** [comment]
@Vectorrent thanks for bringing this up.

> [The official SophiaG implemenation](https://github.com/Liuhong99/Sophia/tree/main) DOES NOT have this problem

SophiaG doesn't have any autograd-related codes, so it's probably not dealing with memory leak issues, unlike SophaiH.

> The main difference between your code and my code has to do with the complexity of the model. Your example model is extremely, extremely simple, while mine is complex - with extremely complex gradients. I believe this is where the true problem lies.

your insight is correct. Also as you mentioned above, SophiaH memory usage should stay consistent.

I believe the problem is all about `torch.autograd.grad` at [here](https://github.com/kozistr/pytorch_optimizer/blob/main/pytorch_optimizer/base/optimizer.py#L104) with complex gradient flows. You can also find lots of memory leak issues with custom autograd function (or custom complex operations) in [here1](https://github.com/pytorch/pytorch/issues/51978), [here2](https://discuss.pytorch.org/t/memory-leakage-in-autograd-grad/143649).

as far as I know, SophiaH needs [this autograd line](https://github.com/kozistr/pytorch_optimizer/blob/main/pytorch_optimizer/base/optimizer.py#L104) to update the hessian (perhaps I might be wrong, you can enlighten me), so I suggest using another optimizer, not using autograd stuff so far.

In short,
1. `autograd` with complex gradient flows is the problem.
2. maybe there's no implementation issue. of course, yours seems fine to me also. (you could try another optimizer not using `autograd` stuff.)
3. I'll take a look at the autograd things later.

thanks!

**Comment by Vectorrent** [comment]
Thanks for the update! It's good to know that I'm not crazy. Sadly, I probably don't have the expertise to debug such a problem currently, but I will try to keep this on my radar - and maybe I'll figure it out, some day.

I find the second-order nature of Sophia to be extremely compelling, and I would love to test with it, at some point.

If I could offer some potential insight: Sophia is supposed to be an _approximation_ of second-order gradients, not the _actual_ second-order gradients. If `autograd` is trying to compute the true second-order grads, instead of an approximation... then yeah, I would expect it to have memory issues.

**Comment by kozistr** [comment]
@Vectorrent 

> If I could offer some potential insight: Sophia is supposed to be an approximation of second-order gradients, not the actual second-order gradients. If autograd is trying to compute the true second-order grads, instead of an approximation... then yeah, I would expect it to have memory issues.

yeap as you said, if we compute the true second-order gradients (full Hessian matrix), then it will need tones of memory not fitted in GPU memory. But, in this case, Hessian Estimator requires only Hessian-vector product instead of the full Hessian matrix, which is memory, and run-time efficient.

I'll mention this issue in the FAQ documentation and close this issue for now. #280.

if you have any questions, feel free to re-open or comment here!

**Comment by kozistr** [comment]
just for the note, https://lightning.ai/forums/t/resolving-backward-with-create-graph/2289
